### DISCUSSION TITLE
Looking for a way to inspect the conclusion postgres makes on whether to permit an action on a table

### DISCUSSION THREAD

**Discussion by paulschneider** [supabase/supabase#30621] [Questions]
I'm having a head scratching couple of days trying to debug why a call to update a database table is failing when initiated from one context but not another. I'm seeking advise on how I can determine what is "wrong" with the failing case.

Some context:

- I’m developing an iOS app using Swift
- I have offloaded all Supabase interactions to a REST API built in NestJS
- I have a user profile context within the app and a corresponding table in Supabase
- I send a JWT to the API which is inspected and from which the Supabase Auth User ID value is extracted (this is the ‘sub’ in the JWT)
- I give this user ID to Supabase via the` .eq()` method of the database call
- RLS is active on the table with a policy that permits read/write/update of a row owned by the authenticated user

The issue is that this workflow works from one context but not another:

- I can successfully use a PUT request to the API to update the profile from a HTTP client (I’m using Insomnia). I get a 200(ok) in response and read out of the profile data as per the 'select' call I make.
- When I make this same API call from within the iOS app everything is apparently successful except that the database update is not committed and a 200(ok) response, but with empty data array, is returned. Other than the update not saved to the database, this empty array is an indicator that the user does not have permission to read the database content  (which itself would make the update fail)

I have tried disabling RLS but the update is again:

- successful from Insomnia
- failing from iOS

Further context

- I use the same JWT in both contexts so rule out security failures with the API
- Tokens are not expired
- Keeping or removing RLS on the table has no effect (always failure from the iOS call)
- Reviewing logs doesn’t show anything abnormal
- Reviewing header values, whilst different between the platforms, doesn’t indicate anything thats wrong

…

Given the platform I have described it’s difficult to share any code that might indicate where a problem lies. I in fact don’t think it’s my code… but there is nothing anywhere to suggest what is causing the update to fail.

Probably needless to say the API is the common denominator between iOS and Insomnia and I am able to successfully:
 
- validate that the user is logged in via the JWT in both context (`.getUser()`)
- set the session with the tokens that have either been supplied or refreshed (`.setSession()`)
- console.log the user object and confirm the authenticated status of the user
- console.log the user ID (it's a UUID) to validate it against the database content

Via the awesome Supabase UI:

- Simulate the update call with the same user (impersonating them) and validate that they have the correct RLS permissions
- Validate the user is in the auth.users table
- ... many other things over the past days.

...

I don't "believe" that this is an RLS issue otherwise the Insomnia call would also fail; I wen't down a path of believing it might be the comparison of the user_id in the `profiles` table Vs the uID in the `auth.users ` and ended up with a policy which I believe casts the user_id value to a UUID:

```postgres
alter policy "Users can update their own data"
on "public"."userProfiles"
to authenticated
using (
    (( SELECT auth.uid() AS uid) = (user_id)::uuid)
with check (
   (( SELECT auth.uid() AS uid) = (user_id)::uuid)
);
```

...

I also don't think this is a bug in Supabase, but I've run out of things to search around and test. Ideally I would love to come across a button that shows me how Postgres / Supabase is making a decision that the current authenticated user is not permitted to carry out the action they are trying to take. This would then lead me to finding out what is wrong... Does this magic button or operation exist?

Final point; I was able to copy and paste the update call that is failing and refactor it into a `SELECT` operation (I.e Get User). This uses exactly the same workflow of JWT introspection, validation and session assurance, so:

```javascript
const response = await Client.connection
        .from(this.tableName)
        .update({
          "given_name": updateProfileDto.givenName,
          "family_name": updateProfileDto.familyName,
        })
        .eq("user_id", this.session.getUser().id)
        .select()
```

Became this:

```javascript
const response = await Client.connection
        .from(this.tableName)
        .select("family_name, given_name, user_id")
        .eq("user_id", this.session.getUser().id)
        .select()
``` 

... and it worked fine from iOS and Insomnia. To me this rules out iOS as a factor completely, and illustrated the workflow is sound and operable. But the `UPDATE` never works.

_[everything tells me that the user has permission to do what they need and every test I have run confirms it. Nothing tells me (that I can find) why this isn't true]_

**Comment by GaryAustin1** [comment]
Just skimming, but what does disable RLS mean?   If you turn it off on the table then the only reason for a 200 on update is you don't match the filter.  If you just turned off update policy, you also have to turn of select policy as update needs both. 

Your last test though seems to indicate select RLS is not involved and filter is the same...

Are you sure your iOS app is going to the same instance as your Insomnia app?

Check the dashboard API gateway logs an insure requests are getting to the instance.

Do you have any trigger functions involved on update?

**Reply by paulschneider** [reply]
Thanks for these thoughts @GaryAustin1 :)

I've been doing all I can to work out what is happening and one of the re-occurring themes when people have issues storing/updating/selecting with postgres is when Row Level Security is enabled but doesn't have an appropriate policy enabled. I have three (policies) and so I tested temporarily disabling the security entirely and seeing if that made a difference. It didn't and I re-enabled RLS.

The policies cover read, update and insert. 

_(that I can make a successful call via Insomnia indicates that the policies, when RLS is enabled, are appropriate and provides the correct permissions)._

I don't believe RLS is the culprit (entirely). And yes the filter on query for Update and Select (PUT & GET) are the same. Since I wrote the above I have also implemented an Insert (POST) and this also works as expected (I.e I can insert data into the database via the iOS and Insomnia apps).

I spent a good deal of time yesterday comparing request data in the logs and didn't find anything that indicated a mismatch when updates are called. Given these tests, and the success I see in `Select` and `Insert` I can guarantee the iOS app is talking to the same API (and therefore SupaBase) instance as Insomnia.

To my knowledge there are no trigger functions involved, at least none that I have implemented (as the sole developer of this platform).

That Insert works is now even more puzzling!

**Reply by GaryAustin1** [reply]
You did not confirm the Gateway API log.
You might try and add an update trigger to a plpgsql function and add `raise log 'new = %',new;` after Begin.  Then look in the Postgres log to see that it occurs and what your columns are.

I don't know of a case of wrong table and columns not erroring (you have a var for your table).   It surely has to be something odd as there is no way update does not work.

You might restart your instance but I don't see how that would help.

I'm out for the night.  Good luck.

**Reply by paulschneider** [reply]
Thanks @GaryAustin1 

Sorry I meant to confirm that the log investigation I had carried out was around the API gateway AND Postgres logs. There's nothing in there to suggest a failure of any kind.

I would also restate that I don't believe for a second that Supabase is failing, just that I'm not easily able to debug (with my current knowledge levels on the matter) WHY my update is failing when other actions are successful under the same circumstances.

This has gotten me a little further. Not being quite sure how to implement logging the way you suggest I do now have a trigger on all updates to the profiles table for any action..... Updating via Insomnia results in a log entry for the update being written, where as the iOS path does not result in an update log entry.

I'll continue to investigate.

Have a great night! 🍻 

_Edit: Though the fact that the update is now confirmed to not fire for me is obvious because the changes aren't committed. I still need to validate why the auth is failing for me. Dang._

**Reply by paulschneider** [reply]
I finally got to the bottom of this, and of course the error was mine rather than any system failing or security mis-configuration....

There were in fact several issues:

1.) In Swift there is concept of 'CodingKeys' which is a way for Swift to understand how to encode and decode JSON data into appropriately typed objects, without telling Swift exactly what we are sending or expecting it doesn't know what to do. Being new to swift I had implemented these keys against the payload object for the update request which turned (E.g) givenName into given_name.

2.) Being new to Nest I had extended a validation DTO from a base class (createUser) which did not do what I expected. Therefore in UpdateUser this effectively passed along undefined values to the updateUser() function.

3.) Having undefined values in the request data resulted in asking postgres to set the (e.g) givenName to undefined which it silently refused to do.

Only a series of logging uncovered this . The request travelling from the client iOS app all the way through to the database call to update() led down the path of assuming this was a permissions issue with the table because it all worked but never committed. There were no messages from Supabase to indicate what the problem was and I headed down the rabbit hole of Googling and ChatGPT to work out how to get to the bottom of it.
